src/HealthChecks/HealthChecks/src/DefaultHealthCheckService.cs

```diff
@@ -158,15 +158,26 @@ private async Task<HealthReportEntry> RunCheckAsync(IServiceScope scope, HealthC
         private static void ValidateRegistrations(IEnumerable<HealthCheckRegistration> registrations)
         {
             // Scan the list for duplicate names to provide a better error if there are duplicates.
-            var duplicateNames = registrations
-                .GroupBy(c => c.Name, StringComparer.OrdinalIgnoreCase)
-                .Where(g => g.Count() > 1)
-                .Select(g => g.Key)
-                .ToList();
 
-            if (duplicateNames.Count > 0)
+            StringBuilder? builder = null;
+            var distinctRegistrations = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            foreach (var registration in registrations)
+            {
+                if (!distinctRegistrations.Add(registration.Name))
+                {
+                    if (builder is null)
+                    {
+                        builder = new StringBuilder("Duplicate health checks were registered with the name(s): ");
+                    }
```

Comment:
```suggestion
                    builder ??= new StringBuilder("Duplicate health checks were registered with the name(s): ");
```

Reply:
@pranavkm, can you estimate how many times this code runs daily throughout the world?

It's not just about making each request faster. It's about using less resources overall, which alleviates energy consumption and carbon emissions.

https://docs.microsoft.com/en-us/learn/modules/sustainable-software-engineering-overview/

Reply:
(builder ??= new StringBuilder("Duplicate health checks were registered with the name(s): "))
        .Append(registration.Name).Append(", ");

😄